### PR TITLE
Add Business ID in Patient Edit Screen

### PR DESCRIPTION
**Pull request by rakshakhegde**
https://www.pivotaltracker.com/story/show/168980840

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt [hunk 1]
```diff
@@ -54,6 +59,19 @@ class EditPatientEffectHandler @AssistedInject constructor(
...
+  private fun fetchAllPrefillForm(scheduler: Scheduler) =
+      ObservableTransformer<FetchBangladeshNationalIdEffect, EditPatientEvent> { effectStream ->
+        effectStream
+            .flatMap { patientRepository.bangladeshNationalIdForPatient(it.patient.uuid).subscribeOn(scheduler) }
+            .map {
+              when (it) {
+                is Just -> it.value.identifier.value
+                None -> ""
+              }
+            }
```
Comment (sanchita) [inline app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt]: There is already a utility function called `unwrapJust` which just gives you the `Just` from an `Optional`.
You won't need this `when` block.

File: app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt [hunk 2]
```diff
@@ -63,12 +63,7 @@ class EditPatientEffectHandler @AssistedInject constructor(
       ObservableTransformer<FetchBangladeshNationalIdEffect, EditPatientEvent> { effectStream ->
         effectStream
             .flatMap { patientRepository.bangladeshNationalIdForPatient(it.patient.uuid).subscribeOn(scheduler) }
-            .map {
-              when (it) {
-                is Just -> it.value.identifier.value
-                None -> ""
-              }
-            }
+            .filterAndUnwrapJust()
```
Comment (sanchita) [inline app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt]: Yes. This is what I was talking about :)

File: app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt [hunk 3]
```diff
@@ -172,6 +186,18 @@ class EditPatientEffectHandler @AssistedInject constructor(
...
 
+  private fun saveBangladeshNationalId(savePatientEffects: Observable<SavePatientEffect>): Observable<EditPatientEvent> {
+    return savePatientEffects
+        .flatMapCompletable {
+          if (it.bangladeshNationalId != null) {
+            patientRepository.saveBusinessId(it.bangladeshNationalId)
+          } else {
+            Completable.complete()
+          }
+        }
```
Comment (sanchita) [inline app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt]: I think we can replace this `if` condition to use a `filter`?

File: app/src/main/res/layout/screen_edit_patient.xml [hunk 4]
```diff
@@ -278,6 +311,38 @@
...
+            android:inputType="textPostalAddress|textCapSentences" />
+        </com.google.android.material.textfield.TextInputLayout>
+
+        <org.simple.clinic.newentry.clearbutton.ClearFieldImageButton
+          style="@style/Clinic.Button.ClearEditText"
+          android:layout_width="wrap_content"
+          android:layout_height="wrap_content"
+          android:layout_gravity="end"
+          app:clearableField="@id/zoneEditText" />
+      </FrameLayout>
```
[the remark this answers is not in the document]
Comment (rakshakhegde) [reply app/src/main/res/layout/screen_edit_patient.xml]: Making these changes now seem to be a huge change. This has to anyways go along with Address Models so let's keep these?

File: app/src/main/java/org/simple/clinic/editpatient/EditPatientEvent.kt [hunk 5]
```diff
@@ -52,3 +52,7 @@ object PatientSaved : EditPatientEvent() {
 object BackClicked: EditPatientEvent() {
   override val analyticsName = "Edit Patient Entry:Back Clicked"
 }
+
+data class NationalIdPrefilled(val bangladeshNationalId: String) : EditPatientEvent() {
+  override val analyticsName: String = "Edit Patient Entry: Prefill form details fetched"
```
[the remark this answers is not in the document]
Comment (rakshakhegde) [reply app/src/main/java/org/simple/clinic/editpatient/EditPatientEvent.kt]: Done.

File: app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt (hunk 1 above)
Comment (rakshakhegde) [reply app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt]: This change has been done in the following commits.

File: app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt (hunk 2 above)
Comment (rakshakhegde) [reply app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt]: Cool :)

File: app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt (hunk 3 above)
Comment (rakshakhegde) [reply app/src/main/java/org/simple/clinic/editpatient/EditPatientEffectHandler.kt]: Done.

**Pull request merged**